lib/services/local/inventory.go

```diff
@@ -0,0 +1,149 @@
+			}
+			return stream.Fail[types.Instance](trace.Wrap(err))
+		}
+		if !req.Match(instance) {
+			return stream.Empty[types.Instance]()
+		}
+		return stream.Once(instance)
+	}
+
+	startKey := backend.Key(instancePrefix, "")
+	endKey := backend.RangeEnd(startKey)
+	items := backend.StreamRange(ctx, s, startKey, endKey, pageSize)
+	return stream.FilterMap(items, func(item backend.Item) (types.Instance, bool) {
+		var instance types.InstanceV1
+		if err := utils.FastUnmarshal(item.Value, &instance); err != nil {
+			s.log.Warnf("Skipping instance at %s, failed to unmarshal: %v", item.Key, err)
+			return nil, false
+		}
+		if err := instance.CheckAndSetDefaults(); err != nil {
+			s.log.Warnf("Skipping instance at %s: %v", item.Key, err)
+			return nil, false
+		}
+		if !req.Match(&instance) {
+			return nil, false
+		}
+		return &instance, true
+	})
+}
+
+// GetRawInstance gets an instance resource by server ID.
+func (s *PresenceService) GetRawInstance(ctx context.Context, serverID string) (types.Instance, []byte, error) {
+	item, err := s.Get(ctx, backend.Key(instancePrefix, serverID))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, nil, trace.NotFound("failed to locate instance %q", serverID)
+		}
+		return nil, nil, trace.Wrap(err)
+	}
+
+	var instance types.InstanceV1
+	if err := utils.FastUnmarshal(item.Value, &instance); err != nil {
+		return nil, nil, trace.BadParameter("failed to unmarshal instance %q: %v", serverID, err)
+	}
+
+	if err := instance.CheckAndSetDefaults(); err != nil {
+		return nil, nil, trace.BadParameter("instance %q appears malformed: %v", serverID, err)
+	}
+
+	return &instance, item.Value, nil
+}
+
+// CompareAndSwapInstance creates or updates the underlying instance resource based on the currently
+// expected value. The first call to this method should use the value returned by GetRawInstance for the
+// 'expect' parameter. Subsequent calls should use the value returned by this method.
+func (s *PresenceService) CompareAndSwapInstance(ctx context.Context, instance types.Instance, expect []byte) ([]byte, error) {
+	if err := instance.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// instance resource expiry is calculated relative to LastSeen and/or the longest living
+	// control log entry (whichever is further in the future).
+	if instance.GetLastSeen().IsZero() || instance.Expiry().IsZero() {
+		instance.SetLastSeen(s.Clock().Now().UTC())
+		instance.SyncLogAndResourceExpiry(apidefaults.ServerAnnounceTTL)
+	}
+
+	v1, ok := instance.(*types.InstanceV1)
+	if !ok {
+		return nil, trace.BadParameter("unexpected type %T, expected %T", instance, v1)
+	}
+
+	value, err := utils.FastMarshal(v1)
+	if err != nil {
+		return nil, trace.Errorf("failed to marshal Instance: %v", err)
+	}
+
+	item := backend.Item{
+		Key:     backend.Key(instancePrefix, instance.GetName()),
+		Value:   value,
+		Expires: instance.Expiry(),
+	}
+
+	if len(expect) == 0 {
```

Comment:
I'm not sure if I like this - we don't allow swapping in an empty bytestring to delete a value in CAS operations, for instance.

Reply:
Less consistent w/ how `Backend` work, but this pattern definitely feels intuitive from a usage-perspective.  `lastHeartbeat == nil` implying "this is first heartbeat" is pretty pervasive, and I feel like doing this makes the code feel very consistent.  IDK.  Not married to it, but I like how well this fits w/ the heartbeat and testing logic.